File: src/yoinkc/renderers/merge_notes.py
"""Render merge-notes.md — fleet merge ambiguity drill-down."""

import hashlib
from pathlib import Path
from typing import NamedTuple

from ..schema import InspectionSnapshot


class _VariantInfo(NamedTuple):
    path: str
    item_type: str
    fleet_count: int
    fleet_total: int
    content_hash: str
    is_winner: bool
    is_tie: bool


def _collect_variant_items(snapshot: InspectionSnapshot) -> list[_VariantInfo]:
    """Collect all non-unanimous variant items across all five item types."""
    items: list[_VariantInfo] = []

    def _add_items(item_list, item_type: str, content_fn):
        if not item_list:
            return
        groups: dict[str, list] = {}
        for item in item_list:
            groups.setdefault(item.path, []).append(item)
        for path, variants in groups.items():
            if not any(v.fleet for v in variants):
                continue
            total = variants[0].fleet.total if variants[0].fleet else 0
            if total == 0:
                continue
            if len(variants) == 1 and variants[0].fleet.count == total:
                continue
            for v in variants:
                c_hash = hashlib.sha256(
                    content_fn(v).encode()
                ).hexdigest()  # Full 64-char SHA-256
                items.append(_VariantInfo(
                    path=v.path,
                    item_type=item_type,
                    fleet_count=v.fleet.count if v.fleet else 0,
                    fleet_total=total,
                    content_hash=c_hash,
                    is_winner=v.include,
                    is_tie=getattr(v, "tie", False),
                ))

    if snapshot.config and snapshot.config.files:
        _add_items(snapshot.config.files, "config", lambda v: v.content)
    if snapshot.services and snapshot.services.drop_ins:
        _add_items(snapshot.services.drop_ins, "drop-in", lambda v: v.content)
    if snapshot.containers:
        if snapshot.containers.quadlet_units:
            _add_items(snapshot.containers.quadlet_units, "quadlet", lambda v: v.content)
        if snapshot.containers.compose_files:
            _add_items(
                snapshot.containers.compose_files, "compose",
                lambda v: str(sorted((img.service, img.image) for img in v.images)),
            )
    if snapshot.non_rpm_software and snapshot.non_rpm_software.env_files:
        _add_items(snapshot.non_rpm_software.env_files, "env", lambda v: v.content)

    return items


def render_merge_notes(snapshot: InspectionSnapshot, output_dir: Path) -> None:
    """Write merge-notes.md if there are non-unanimous or tied items."""
    items = _collect_variant_items(snapshot)
    if not items:
        return

    lines = [
        "# Fleet Merge Notes",
        "",
        "This file documents fleet merge decisions where hosts disagreed on file content.",
        "Review these items to verify the auto-selected variant is correct for your target image.",
        "",
    ]

    tied: dict[str, list[_VariantInfo]] = {}
    non_unanimous: dict[str, list[_VariantInfo]] = {}

    for item in items:
        if item.is_tie:
            tied.setdefault(item.path, []).append(item)
        else:
            non_unanimous.setdefault(item.path, []).append(item)

    if tied:
        lines.append("## Tied Items (auto-resolved by content hash)")
        lines.append("")
        for path, variants in sorted(tied.items()):
            item_type = variants[0].item_type
            total = variants[0].fleet_total
            winner = next((v for v in variants if v.is_winner), None)
            lines.append(f"### `{path}` ({item_type})")
            lines.append("")
            lines.append(f"- **Variants:** {len(variants)}")
            lines.append(f"- **Fleet total:** {total} hosts")
            if winner:
                lines.append(f"- **Auto-selected:** hash `{winner.content_hash}` ({winner.fleet_count}/{total} hosts)")
            lines.append("")
            lines.append("| Variant hash | Hosts | Selected |")
            lines.append("|---|---|---|")
            for v in sorted(variants, key=lambda x: x.content_hash):
                selected = "**winner**" if v.is_winner else "—"
                lines.append(f"| `{v.content_hash}` | {v.fleet_count}/{v.fleet_total} | {selected} |")
            lines.append("")

    if non_unanimous:
        lines.append("## Non-Unanimous Items")
        lines.append("")
        lines.append("These items have a clear winner but were not present on all hosts.")
        lines.append("")
        for path, variants in sorted(non_unanimous.items()):
            item_type = variants[0].item_type
            total = variants[0].fleet_total
            winner = next((v for v in variants if v.is_winner), None)
            if winner:
                lines.append(f"- `{path}` ({item_type}): winner at {winner.fleet_count}/{total} hosts, "
                             f"{len(variants)} variant(s)")
            else:
                lines.append(f"- `{path}` ({item_type}): {len(variants)} variant(s), no winner selected")

    output_dir.mkdir(parents=True, exist_ok=True)
    (output_dir / "merge-notes.md").write_text("\n".join(lines) + "\n")

File: cmd/inspectah/internal/renderer/mergenotes.go
package renderer

import (
	"crypto/sha256"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/marrusl/inspectah/cmd/inspectah/internal/schema"
)

// variantInfo holds data about a fleet variant item.
type variantInfo struct {
	path        string
	itemType    string
	contentHash string
	fleetCount  int
	fleetTotal  int
	isWinner    bool
	isTie       bool
}

// RenderMergeNotes produces merge-notes.md documenting fleet merge
// decisions where hosts disagreed on file content.
func RenderMergeNotes(snap *schema.InspectionSnapshot, outputDir string) error {
	items := collectVariantItems(snap)
	if len(items) == 0 {
		return nil // No fleet data = no merge-notes.md
	}

	var lines []string
	lines = append(lines, "# Fleet Merge Notes")
	lines = append(lines, "")
	lines = append(lines, "This file documents fleet merge decisions where hosts disagreed on file content.")
	lines = append(lines, "Review these items to verify the auto-selected variant is correct for your target image.")
	lines = append(lines, "")

	// Group by path
	tied := make(map[string][]variantInfo)
	nonUnanimous := make(map[string][]variantInfo)

	for _, item := range items {
		path := item.path
		if item.isTie {
			tied[path] = append(tied[path], item)
		} else if item.fleetCount < item.fleetTotal {
			nonUnanimous[path] = append(nonUnanimous[path], item)
		}
	}

	if len(tied) > 0 {
		lines = append(lines, "## Tied Items (auto-resolved by content hash)")
		lines = append(lines, "")

		// Sort paths for deterministic output
		paths := make([]string, 0, len(tied))
		for p := range tied {
			paths = append(paths, p)
		}
		sort.Strings(paths)

		for _, path := range paths {
			variants := tied[path]
			if len(variants) == 0 {
				continue
			}
			itemType := variants[0].itemType
			total := variants[0].fleetTotal
			var winner *variantInfo
			for i := range variants {
				if variants[i].isWinner {
					winner = &variants[i]
					break
				}
			}

			lines = append(lines, fmt.Sprintf("### `%s` (%s)", path, itemType))
			lines = append(lines, "")
			lines = append(lines, fmt.Sprintf("- **Variants:** %d", len(variants)))
			lines = append(lines, fmt.Sprintf("- **Fleet total:** %d hosts", total))
			if winner != nil {
				lines = append(lines, fmt.Sprintf("- **Auto-selected:** hash `%s` (%d/%d hosts)",
					winner.contentHash, winner.fleetCount, total))
			}
			lines = append(lines, "")
			lines = append(lines, "| Variant hash | Hosts | Selected |")
			lines = append(lines, "|---|---|---|")
			sort.Slice(variants, func(i, j int) bool {
				return variants[i].contentHash < variants[j].contentHash
			})
			for _, v := range variants {
				selected := "--"
				if v.isWinner {
					selected = "**winner**"
				}
				lines = append(lines, fmt.Sprintf("| `%s` | %d/%d | %s |",
					v.contentHash, v.fleetCount, v.fleetTotal, selected))
			}
			lines = append(lines, "")
		}
	}

	if len(nonUnanimous) > 0 {
		lines = append(lines, "## Non-Unanimous Items")
		lines = append(lines, "")
		lines = append(lines, "These items have a clear winner but were not present on all hosts.")
		lines = append(lines, "")

		paths := make([]string, 0, len(nonUnanimous))
		for p := range nonUnanimous {
			paths = append(paths, p)
		}
		sort.Strings(paths)

		for _, path := range paths {
			variants := nonUnanimous[path]
			if len(variants) == 0 {
				continue
			}
			itemType := variants[0].itemType
			total := variants[0].fleetTotal

			var winner *variantInfo
			for i := range variants {
				if variants[i].isWinner {
					winner = &variants[i]
					break
				}
			}
			if winner != nil {
				lines = append(lines, fmt.Sprintf("- `%s` (%s): winner at %d/%d hosts, %d variant(s)",
					path, itemType, winner.fleetCount, total, len(variants)))
			} else {
				lines = append(lines, fmt.Sprintf("- `%s` (%s): %d variant(s), no winner selected",
					path, itemType, len(variants)))
			}
		}
	}

	content := strings.Join(lines, "\n") + "\n"
	return os.WriteFile(filepath.Join(outputDir, "merge-notes.md"), []byte(content), 0644)
}

// collectVariantItems extracts items with fleet prevalence data for merge notes.
func collectVariantItems(snap *schema.InspectionSnapshot) []variantInfo {
	var items []variantInfo

	// Config files
	if snap.Config != nil {
		for _, f := range snap.Config.Files {
			if f.Fleet == nil {
				continue
			}
			items = append(items, variantInfo{
				path:        f.Path,
				itemType:    "config",
				contentHash: contentHash(f.Content),
				fleetCount:  f.Fleet.Count,
				fleetTotal:  f.Fleet.Total,
				isWinner:    f.TieWinner,
				isTie:       f.Tie,
			})
		}
	}

	// Drop-ins
	if snap.Services != nil {
		for _, d := range snap.Services.DropIns {
			if d.Fleet == nil {
				continue
			}
			items = append(items, variantInfo{
				path:        d.Path,
				itemType:    "drop-in",
				contentHash: contentHash(d.Content),
				fleetCount:  d.Fleet.Count,
				fleetTotal:  d.Fleet.Total,
				isWinner:    d.TieWinner,
				isTie:       d.Tie,
			})
		}
	}

	// Quadlet units
	if snap.Containers != nil {
		for _, q := range snap.Containers.QuadletUnits {
			if q.Fleet == nil {
				continue
			}
			items = append(items, variantInfo{
				path:        q.Path,
				itemType:    "quadlet",
				contentHash: contentHash(q.Content),
				fleetCount:  q.Fleet.Count,
				fleetTotal:  q.Fleet.Total,
				isWinner:    q.TieWinner,
				isTie:       q.Tie,
			})
		}

		for _, c := range snap.Containers.ComposeFiles {
			if c.Fleet == nil {
				continue
			}
			items = append(items, variantInfo{
				path:        c.Path,
				itemType:    "compose",
				contentHash: fmt.Sprintf("%x", sha256.Sum256([]byte(c.Path)))[:8],
				fleetCount:  c.Fleet.Count,
				fleetTotal:  c.Fleet.Total,
				isWinner:    c.TieWinner,
				isTie:       c.Tie,
			})
		}
	}

	// Non-RPM env files
	if snap.NonRpmSoftware != nil {
		for _, e := range snap.NonRpmSoftware.EnvFiles {
			if e.Fleet == nil {
				continue
			}
			items = append(items, variantInfo{
				path:        e.Path,
				itemType:    "env",
				contentHash: contentHash(e.Content),
				fleetCount:  e.Fleet.Count,
				fleetTotal:  e.Fleet.Total,
				isWinner:    e.TieWinner,
				isTie:       e.Tie,
			})
		}
	}

	return items
}

// contentHash returns a short SHA-256 hash of content, matching the
// Python merge normalization.
func contentHash(content string) string {
	// Normalize: strip trailing whitespace per line, strip leading/trailing blank lines
	lines := strings.Split(content, "\n")
	var normalized []string
	for _, l := range lines {
		normalized = append(normalized, strings.TrimRight(l, " \t\r"))
	}
	text := strings.TrimSpace(strings.Join(normalized, "\n"))
	hash := sha256.Sum256([]byte(text))
	return fmt.Sprintf("%x", hash)[:8]
}
